cpp/src/arrow/io/file.cc

```diff
@@ -378,6 +378,106 @@ Status FileOutputStream::Write(const void* data, int64_t length) {
 
 int FileOutputStream::file_descriptor() const { return impl_->fd(); }
 
+// ----------------------------------------------------------------------
+// DirectFileOutputStream, change the Open, Write and Close methods from FileOutputStream
+// Uses DirectIO for writes. Will only write out things in disk-sector-size byte blocks.
+// Buffers leftover bytes in an internal data structure, which will be padded to
+// disk-sector-size bytes and written upon call to close.
+
+#if defined(__linux__)
+class DirectFileOutputStream::DirectFileOutputStreamImpl : public OSFile {
+ public:
+  Status Open(const std::string& path) {
+    return OpenWritable(path, true, false, true /* write_only */, true);
+  }
+  Status Open(int fd) { return OpenWritable(fd); }
+};
+
+DirectFileOutputStream::DirectFileOutputStream(int64_t sector_size) {
+  sector_size_ = sector_size;
+  uintptr_t mask = (uintptr_t)(sector_size - 1);
```

Comment:
Please use C++ casts, not C, so e.g. `static_cast`.

Reply:
Also can we check that `sector_size` is a power of two?

Reply:
I can check if the sector_size is a power of two. But a constructor can't return a Result right, what's the right way to do this in Arrow? I don't think just adding this code:
```
if !(sector_size != 0 && (sector_size & (sector_size-1)) == 0)
  {
    return Status::IOError("Length must be non-negative");
  }
```
is going to work here because return Status shouldn't be allowed in constructor.